### PR TITLE
use groovy to generate class & use static method to init model map

### PR DESCRIPTION
**Pull request by runzexia**
Signed-off-by: runzexia <runzexia@yunify.com>

### CODE REVIEW

**Comment by runzexia** [comment]
@gavinfish @MyronFanQiuIn 
In this PR, I used groovy to generate a java file and initialize the map myself. This is a draft. I want to know if you think it is a solution that can be accepted.

**Comment by gavinfish** [comment]
I think this could be an option here. Could you please fix the CI job first, thanks.

**Comment by runzexia** [comment]
done

File: pom.xml [hunk 1]
```diff
@@ -155,16 +175,64 @@
             <plugin>
                 <groupId>org.jenkins-ci.tools</groupId>
                 <artifactId>maven-hpi-plugin</artifactId>
+                <version>3.6</version>
```
Comment (jglick) [inline pom.xml]:
```suggestion
```

Unnecessary—this is already the default in the parent POM.

File: pom.xml [hunk 2]
```diff
@@ -155,16 +175,64 @@
...
                         com.fasterxml.jackson.
                         com.microsoft.jenkins.azurecommons.core.
+                        com.google.
                     </maskClasses>
                 </configuration>
             </plugin>
+            <plugin>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <source>1.8</source>
```
Comment (jglick) [inline pom.xml]: Can be deleted. The parent POM already does this.

File: src/main/java/com/microsoft/jenkins/kubernetes/wrapper/V1ResourceManager.java [hunk 3]
```diff
@@ -21,17 +22,35 @@
...
+           STRATEGIC_MERGE_PATCH_CLIENT = ClientBuilder.standard().setOverridePatchFormat(
+                   ApiClient.PATCH_FORMAT_STRATEGIC_MERGE_PATCH).build();
+        } catch (IOException e) {
+            throw new RuntimeException(e.getMessage());
+        }
+
+    }
+
+    private static final ApiClient STRATEGIC_MERGE_PATCH_CLIENT;
     private static final CoreV1Api CORE_V1_API_INSTANCE = new CoreV1Api();
```
Comment (runzexia) [inline src/main/java/com/microsoft/jenkins/kubernetes/wrapper/V1ResourceManager.java]: I found that the client here is a global client. When several tasks are running at the same time, the client may be reset by other tasks. Should I add a client field to the ResourceUpdater?
@gavinfish

**Comment by gavinfish** [comment]
@runzexia Thanks for your contributions here. Is it possible to separate the `generate class` and `init model` in a single PR and I will merge this first as an infrastructure work for migrate the sdk. This PR is really complex now and I think it is not good for later track.

**Comment by runzexia** [comment]
Ok, I will split a separate PR today.

File: src/main/script/KubernetesModelClassesGen.groovy [hunk 4]
```diff
@@ -0,0 +1,58 @@
+import io.kubernetes.client.models.*
+import com.google.common.reflect.ClassPath
+import io.kubernetes.client.util.Yaml
+import org.apache.commons.lang3.tuple.Pair;
+
+class KubernetesModelClassesGen {
```
Comment (runzexia) [inline src/main/script/KubernetesModelClassesGen.groovy]: Generate java code using groovy script

File: src/main/java/com/microsoft/jenkins/kubernetes/wrapper/ResourceUpdaterMap.java [hunk 5]
```diff
@@ -7,42 +7,60 @@
...
+import org.apache.commons.lang3.tuple.Pair;
 
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
 
-public final class ResourceUpdaterMap extends HashMap<Class<?>, Class<? extends ResourceManager.ResourceUpdater>> {
-    private static final Map<Class<?>, Class<? extends ResourceManager.ResourceUpdater>> INSTANCE =
+public final class ResourceUpdaterMap extends HashMap<Class<?>,
+        Pair<Class<? extends ResourceManager>, Class<? extends ResourceManager.ResourceUpdater>>> {
```
Comment (runzexia) [inline src/main/java/com/microsoft/jenkins/kubernetes/wrapper/ResourceUpdaterMap.java]: `Map<Class<?>, Class<? extends ResourceManager.ResourceUpdater>>` =>
`Map<Class<?>,Pair<Class<? extends ResourceManager>, Class<? extends ResourceManager.ResourceUpdater>>>`
If you want to construct a ResourceUpdater, you need to construct the ResourceManager first.

**Comment by runzexia** [comment]
I tried to mask out some extraneous code, and now the PR also contains a small amount of construction logic so that the plugin can perform basic create resource.
Also contains some code to fix ci job.
Is this acceptable to you, if not, I will reprocess it.

**Comment by runzexia** [comment]
Is there time to review the code this week?
Currently I have completed the migration of the current support resources at https://github.com/jenkinsci/kubernetes-cd-plugin/pull/88 and added support for some other resources.
I have tested the new features locally, and everything works fine. If we can merge all the PRs this week, I think we can try to release a beta.

**Pull request merged**